### ISSUE TITLE
S3FilesStore can use a lot of memory

### ISSUE THREAD

**Issue by kmike** [scrapy/scrapy#482]
Hi,

@nramirezuy and me were debugging memory issue with one of the spiders  some time ago, and it seems to be caused by ImagesPipeline + [S3FilesStore](https://github.com/scrapy/scrapy/blob/master/scrapy/contrib/pipeline/files.py#L66). I haven't confirmed that it was the cause of memory issue, this ticket is based solely on reading the source code.

FilesPipeline reads the whole file to memory and then defers the uploading to thread (via `S3FilesStore.persist_file`, passing file contents as bytes). So there could be many files loaded to memory at the same time, and as soon as files are downloaded faster than they are are uploaded to s3, memory usage will grow. This is not unlikely IMHO because s3 is not super-fast. For ImagesPipeline it is worse because it uploads not only the image itself, but also the generated thumbnails.

I think S3FilesStore should persist files to temporary location before uploading them to S3 (at least optionally). This would allow streaming files without storing them in memory.

**Comment by Parth-Vader** [comment]
Change the links to https://github.com/scrapy/scrapy/blob/master/scrapy/pipelines/files.py and https://github.com/scrapy/scrapy/blob/master/scrapy/pipelines/media.py since the ones linked are depriciated.

**Comment by G-Rath** [comment]
Are folks aware this has been reported as a security issue that affects most (if not all) versions of Scrapy? https://nvd.nist.gov/vuln/detail/CVE-2017-14158 & https://github.com/pypa/advisory-database/blob/8b7a4d62a95e8f605e5dfb4e0b4f299e6403dc12/vulns/scrapy/PYSEC-2017-83.yaml

I'm currently trying to determine if the security angle has been fixed, and if not what is required for that since we're got this flagged in one of our applications; I'm happy to help with this as much as I can, but I don't work in Python a lot so suspect I'll be of limited use but I'm happy to be apart of discussions and even dive into code if someone more familiar with the codebase could spare some time to support me 🙂

**Comment by fazalmajid** [comment]
This issue is also flagged as [PYSEC-2017-83](https://osv.dev/vulnerability/PYSEC-2017-83) by [pip-audit](https://github.com/trailofbits/pip-audit).

I would recommend putting a hard configurable limit on the total size of downloads iniitiated by a single top-level request to prevent the kind of DoS attacks mentioned. Keep in mind scrapers like ScraPy are often detested by the sites being scraped (despite being [perfectly legal](https://www.theregister.com/2022/04/19/scraping_public_data_linkedin/)) so deliberate DoS countermeasures are well within the realm of possibility.

**Comment by G-Rath** [comment]
@fazalmajid that sounds like a reasonable solution - is that something you'd be comfortable implementing? As I've said I'm not really a Python developer so can't really take on trying to implement a solution myself 😞

**Comment by fazalmajid** [comment]
No, as I don’t use Scrapy myself and have no idea about the code base. Plus since it’s a project sponsored by a commercial organization, they would certainly have opinions about whether to implement this and how, and I am not about to code “on spec”.

**Comment by genismoreno** [comment]
Hi, is there a plan to fix this? I have seen this vulnerability for a while, but I don't see a clear solution for it. Thanks!

**Comment by kmike** [comment]
Hey. It seems this CVE is popping up everywhere, and can cause some warnings for the users, so we should do something about it.

Let's investigate possible solution. https://nvd.nist.gov/vuln/detail/CVE-2017-14158 tells that the issue is an instance of https://cwe.mitre.org/data/definitions/400.html. Description of CWE-400:

> The software does not properly control the allocation and maintenance of a limited resource, thereby enabling an actor to influence the amount of resources consumed, eventually leading to the exhaustion of available resources. 

> Limited resources include memory, file system storage, database connection pool entries, and CPU. If an attacker can trigger the allocation of these limited resources, **but the number or size of the resources is not controlled**, then the attacker could cause a denial of service that consumes all available resources. 

Scrapy controls the number and size of the resources in the following way:

1. There is DOWNLOAD_MAXSIZE (https://docs.scrapy.org/en/latest/topics/settings.html#download-maxsize) option, to prevent downloading large files.
2. There are various CONCURRENT_REQUESTS options, which allow the amount of parallel downloads (each download then causes an upload to S3).
3. There is SCRAPER_SLOT_MAX_ACTIVE_SIZE, which is a soft limit for total size of all responses being processed by scraper ("While the sum of the sizes of all responses being processed is above this value, Scrapy does not process new requests."). I'm not sure though why is it applied on Scraper level, not on Downloader level. So, it seems this option doesn't have an effect if a request is added to downloader without going through scraper, and it seems requests initiated in MediaPipeline don't go through scraper.

Reducing the amount of RAM, e.g. storing more data to disk, is not a solution for CVE. By doing so we're shifting the resource from "memory usage" to "disk usage", without addressing the security issue. 

It seems that having an option similar to SCRAPER_SLOT_MAX_ACTIVE_SIZE, but which works on Downloader level or on Engine level should solve the problem. I.e. limit (in a soft way - if we're over limit, stop accepting new work) the total byte size of responses being processed by all Scrapy components. What do you think?

**Comment by kmike** [comment]
As for vulnerability, if the files/images pipelines or other components which add files to Downloader directly are not used, then it seems CWE-400 doesn't apply. There are DOWNLOAD_MAXSIZE and CONCURRENT_REQUESTS limits, and they look adequate according to the CWE-400 description.

**Comment by G-Rath** [comment]
@kmike I think that sounds look a good idea

**Comment by G-Rath** [comment]
I've submitted an [update to the advisory on GitHub](https://github.com/github/advisory-database/pull/1667) to reflect that this has not been addressed in 2.8

**Comment by melroy89** [comment]
I indeed got a "Scrapy denial of service vulnerability"  CVE-2017-14158

**Comment by ret2libc** [comment]
Hi! Is there any on-going effort to fix this vulnerability? As this has been opened for a long time, I wonder if the issue is just not that critical from a security standpoint as listed on NVD or if it is the fix that is complex.

@kmike Regarding a solution, I see that the `MediaPipeline` is using the `Engine` to download the object
https://github.com/scrapy/scrapy/blob/96033ce5a7f857942e3c6d488c8aab5b4aa03295/scrapy/pipelines/media.py#L196
And the `Engine` does have a `Slot` class that is tracking the requests... Would it be possible to have `ENGINE_SLOT_MAX_ACTIVE_SIZE` similarly to `SCRAPER_SLOT_MAX_ACTIVE_SIZE`? The thing I'm not sure about is: if I check for the limit in the `_download` method (https://github.com/scrapy/scrapy/blob/96033ce5a7f857942e3c6d488c8aab5b4aa03295/scrapy/core/engine.py#L326) how can I postpone the processing of that request?

Thanks!

**Comment by kmike** [comment]
> Hi! Is there any on-going effort to fix this vulnerability? As this has been opened for a long time, I wonder if the issue is just not that critical from a security standpoint as listed on NVD or if it is the fix that is complex

Hey! I don't think anyone works on it actively now. You're right; maybe we're careless, but it indeed doesn't look that critical. I haven't seen anyone reporting any practical problem related to this issue. Note that the issue is not that Scrapy can use a lot of memory, the issue is that there is no perfect ay to set a limit on the amount of memory Scrapy can use. Tbh, the main motivation to fix it is to make the security notifications to go away - which is probably why there are security notifications in a first place, to motivate maintainers, so it works as intended :)

That said, even if that's not super-critical, it's still an important issue, and a nice improvement to Scrapy behavior.

> Would it be possible to have ENGINE_SLOT_MAX_ACTIVE_SIZE similarly to SCRAPER_SLOT_MAX_ACTIVE_SIZE?

Yes, I think that's the idea.

> how can I postpone the processing of that request?

The limit could be "soft" - you don't postpone processing of that request, you postpone processing of subsequent requests.

**Comment by jpmckinney** [comment]
I'm confused as to why this is even a security vulnerability. Are people habitually allowing third-parties (i.e. potential attackers) to cause crawls to start against arbitrary URLs? I just don't see how an attacker can DOS my system, considering it's (1) not publicly accessible, (2) all crawls are to specific URLs as configured in spiders, and (3) crawls are only initiated by a scheduler controlled by the admin (or scheduled by the admin directly). I would have assumed that most Scrapy usage fit these 3 assumptions. It feels a bit like a CVE about `tail`, because if a user gains access to a server they can run `tail /dev/zero` and exhaust memory...

**Comment by fazalmajid** [comment]
The threat is sites that disapprove of being scraped and that could take deliberate measures to crash the scrapers, i.e. the threat actor is the site being scraped itself.

**Comment by jpmckinney** [comment]
Hmm. Do general-use clients for internet protocols set memory limits? It seems like the same threat exists for anything that allows loading a large payload into memory, e.g. `requests`' `response.json()` method.

**Comment by Gallaecio** [comment]
Assumption 2 is not entirely true for web scraping. URL discovery is often done as part of the crawling. A website detecting some activity as automated could indeed send special responses to a crawler, e.g. point a crawler to special URLs designed to mess with the crawling.

I don’t think this is a likely scenario, and I am not sure how much of a security issue this is, since websites doing something like this could expose themselves legally.

We still want to address the issue, though, because security aside, this can be an actual issue on normal websites.

**Comment by aabraham-fc** [comment]
Hey! Just got the denial of service vulnerability  alert on our scrapy repository. Is there a recommended solution to mitigate risk? We have an image pipeline and s3 file store which seems to be the cause for many here.

Thanks!

**Comment by albertedwardson** [comment]
interesting 

will take a look at this

note to myself: this needs backpressure control, just like sitemap spider
